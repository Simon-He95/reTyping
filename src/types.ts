export interface InsertPatch {
  type: 'insert'
  from: number
  text: string
}

export interface RemovalPatch {
  type: 'removal'
  from: number
  length: number
}

export type Patch = InsertPatch | RemovalPatch
