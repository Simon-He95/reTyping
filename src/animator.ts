import { Patch } from "."

function randomRgb() {
  return `rgb(${Math.floor(Math.random() * 255)}, ${Math.floor(
    Math.random() * 255
  )}, ${Math.floor(Math.random() * 255)})`
}

export function* createAnimator(input: string, patches: Patch[]) {
  let output = input
  let cursor = 0
  for (const patch of patches) {
    if (patch.type === 'insert') {
      cursor = patch.from
      const head = output.slice(0, patch.from)
      const tail = output.slice(patch.from)
      let selection = ''
      for (const chart of patch.text) {

        selection += `<span style="color:${randomRgb()}">${chart}</span>`
        yield { cursor: cursor + selection.length, output: head + selection + tail }
      }
      output = head + patch.text + tail
    }
    else if (patch.type === 'removal') {
      cursor = patch.from - patch.length
      const head = output.slice(0, patch.from - patch.length)
      const tail = output.slice(patch.from)
      let selection = output.slice(patch.from - patch.length, patch.from)
      for (let i = selection.length; i >= 0; i--) {
        yield { cursor: cursor + i, output: head + selection.slice(0, i) + tail }
      }
      output = head + tail

    }
  }
}