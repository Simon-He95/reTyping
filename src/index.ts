export * from './animator'
export * from './types'
export * from './patch'
