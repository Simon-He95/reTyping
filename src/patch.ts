import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'
import type { Patch } from './types'

export function diff(a: string, b: string): Diff[] {
  const differ = new DMP()
  const delta = differ.diff_main(a, b)
  differ.diff_cleanupSemantic(delta)
  return delta
}

export function calculatePatch(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  let index = 0
  for (const change of diffs) {
    if (change[0] === 0) {
      index += change[1].length
      continue
    }
    if (change[0] === -1) {
      patches.push({
        type: 'removal',
        from: index + change[1].length,
        length: change[1].length,
      })
    }
    if (change[0] === 1) {
      patches.push({
        type: 'insert',
        from: index,
        text: change[1],
      })
      index += change[1].length
    }
  }
  return patches
}

export function applyPatches(input: string, patches: Patch[]) {
  let text = input
  for (const patch of patches) {
    if (patch.type === 'insert')
      text = text.slice(0, patch.from) + patch.text + text.slice(patch.from)

    else if (patch.type === 'removal')
      text = text.slice(0, patch.from - patch.length) + text.slice(patch.from)
  }
  return text
}
