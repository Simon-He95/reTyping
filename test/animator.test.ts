import { expect, it, } from 'vitest'
import { calculatePatch, diff } from '../src'
import { createAnimator } from '../src/animator'
import { input, output } from './fixture'

it('animator', () => {
  const delta = diff(input, output)
  const patches = calculatePatch(delta)
  const animator = createAnimator(input, patches)
  expect([...animator]).toMatchSnapshot('animator')
})