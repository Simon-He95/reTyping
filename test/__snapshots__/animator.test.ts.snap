// Vitest Snapshot v1

exports[`animator > animator 1`] = `
[
  {
    "cursor": 61,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one, } from '../src'
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 62,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,t } from '../src'
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 63,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,tw } from '../src'
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 64,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 122,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 121,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('exporte', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 120,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('export', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 119,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('expor', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 118,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('expo', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 117,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('exp', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 116,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('ex', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 115,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('e', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 114,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 115,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('o', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 116,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('on', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 117,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 156,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
   })
})
",
  },
  {
    "cursor": 157,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    })
})
",
  },
  {
    "cursor": 158,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
     })
})
",
  },
  {
    "cursor": 159,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
      })
})
",
  },
  {
    "cursor": 160,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    e  })
})
",
  },
  {
    "cursor": 161,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    ex  })
})
",
  },
  {
    "cursor": 162,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    exp  })
})
",
  },
  {
    "cursor": 163,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expe  })
})
",
  },
  {
    "cursor": 164,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expec  })
})
",
  },
  {
    "cursor": 165,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect  })
})
",
  },
  {
    "cursor": 166,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(  })
})
",
  },
  {
    "cursor": 167,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(t  })
})
",
  },
  {
    "cursor": 168,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(tw  })
})
",
  },
  {
    "cursor": 169,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two  })
})
",
  },
  {
    "cursor": 170,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two)  })
})
",
  },
  {
    "cursor": 171,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).  })
})
",
  },
  {
    "cursor": 172,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).t  })
})
",
  },
  {
    "cursor": 173,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).to  })
})
",
  },
  {
    "cursor": 174,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toE  })
})
",
  },
  {
    "cursor": 175,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEq  })
})
",
  },
  {
    "cursor": 176,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqu  })
})
",
  },
  {
    "cursor": 177,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqua  })
})
",
  },
  {
    "cursor": 178,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual  })
})
",
  },
  {
    "cursor": 179,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(  })
})
",
  },
  {
    "cursor": 180,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2  })
})
",
  },
  {
    "cursor": 181,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)  })
})
",
  },
  {
    "cursor": 182,
    "output": "
import { describe, expect, it } from \\"vitest\\";
import { one,two } from '../src'
describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
]
`;
